### PR TITLE
Add to PCMCI Testing

### PR DESCRIPTION
**Pull request by ewengillies**
Test the new functions added in #13, extend tests to cover more use cases.  Can only be merged in after #13.

### CODE REVIEW

**Comment by ewengillies** [comment]
This merge request has several main goals: 
   * Fixing potentially dangerous code in `pcmci.py`
   * Improving the existing testing in `pcmci.py`
   * Adding in new testing for `pcmci.py`

# Fixing Code in `pcmci.py`

There were a few changes made to `pcmci.py`.  They fall into the following catagories:
  * Creating internal copies of mutable arguments
  * Ensuring the condition dimensions have the correct limits
  * Other small checks or changes

## Creating Internal Copies of mutable arguments

Previous, mutable arguments (like dicts or lists) were passed into functions and then changed once inside the function.  This can lead to unexpected behaviour, whereby the user calls the same function twice with the same arguments, but the first call has changed one of the arguments, resulting in different functionality on the second call.  

This was the case for several functions, where now `copy.deepcopy` is used to establish an internal copy of the mutable argument, which is only defined and used in the scope of the function:
  * [_set_selected_variables()](https://github.com/ewengillies/tigramite/blob/58ac591c1cf88b08a10b22516547ec850f3ff168/tigramite/pcmci.py#L268)
  * [_set_sel_links()](https://github.com/ewengillies/tigramite/blob/58ac591c1cf88b08a10b22516547ec850f3ff168/tigramite/pcmci.py#L296)
 * [_get_int_parents()](https://github.com/ewengillies/tigramite/blob/d0dab7f500593a9a449aa9beb572b419fca255c0/tigramite/pcmci.py#L1019)
  * [_int_pc_alpha](https://github.com/ewengillies/tigramite/blob/d0dab7f500593a9a449aa9beb572b419fca255c0/tigramite/pcmci.py#L826), which makes an internal copy of `pc_alpha` in case it is a list.

## Ensuring Correct Condition Dimensions 

Perviously, there were two issues with the conditions dimensions:
  * The condition dimensions passed through the `for` loop in `run_pc_stable_single` was too large at `max_conds_dim + 1`.
  * The condition dimensions ignored `tau_min` completely. 

### Fixing the dimensions of the `for` loop in `run_pc_stable_single`

[This is the `for` loop that this section discusses.](https://github.com/ewengillies/tigramite/blob/58ac591c1cf88b08a10b22516547ec850f3ff168/tigramite/pcmci.py#L593)

The condition dimension now passed through the for loop is `max_conds_dim`, not `max_conds_dim + 1`.  This `+1` existed to extend the `range` function to include `max_conds_dim` dimension itself.  However, the parent itself is always excluded from the "pool" of conditions.  This brings the maximum possible dimensionality down by 1.

Previously, this extra `+1` was benign.  It resulted in code that iterated through the results of non-nonsensical combinatorics like "n choose n+1", which yielded no results to iterate through, and were hence just skipped.  This has now been fixed to the correct limits

### Taking `tau_min` into account

Previously, setting the default `max_conds_dim` did not take tau min into account.  It now does, which results in previously failing tests to now pass.  [The function `_set_max_condition_dim()` carries out this change.](https://github.com/ewengillies/tigramite/blob/58ac591c1cf88b08a10b22516547ec850f3ff168/tigramite/pcmci.py#L750)

## Other small fixes

Other small checks and fixes have been included:
  * Check in [_set_sel_links()](https://github.com/ewengillies/tigramite/blob/58ac591c1cf88b08a10b22516547ec850f3ff168/tigramite/pcmci.py#L296) to ensure selected links are reasonable.
  * Wrapper function called [print_results()](https://github.com/ewengillies/tigramite/blob/58ac591c1cf88b08a10b22516547ec850f3ff168/tigramite/pcmci.py#L1426) prints the results of `run_pcmci` directly from this function's output.
  * [return_significant_parents()](https://github.com/ewengillies/tigramite/blob/58ac591c1cf88b08a10b22516547ec850f3ff168/tigramite/pcmci.py#L1328) is no longer a private function.
  *  [print_significant_links()](https://github.com/ewengillies/tigramite/blob/58ac591c1cf88b08a10b22516547ec850f3ff168/tigramite/pcmci.py#L1371) is no longer a private function.

 # Testing for `pcmci.py`

Testing for `pcmci.py` has now been moved into two separate categories:
  * Ensuring `pcmci.py` comes to the correct answers, introduced in #12 
  * Ensuring all `pcmci.py` constructors and tools yield the right results, introduced here.

## `pcmci.py` Calculation Testing Improvments

The testing that existed for `pcmci.py` from #12 has been moved to a new suite, `test_pcmci_calculation.py`

Previous to this merge request, only one link structure was used for all tests checking that the `pcmci.py` returned the correct.  This chain structure has been parameterized in function [a_chain()](https://github.com/ewengillies/tigramite/blob/9a639565434c76dad9296fac43ceb3a213f9bb96/tests/test_pcmci_calculations.py#L61), such that a variable length chain of variable auto-correlation and causal strength can be defined.  

While only one such generating function is currently defined, others can be defined and used in an identical fashion to improve the range of cases that this testing covers.

## `pcmci.py` Construction and Tool Testing

`pcmci.py` also offers the user many tools and uses many internal functions to carry out its calculations. These functions must also be tested.  To do so, a new testing suite has been added `test_pcmci_construction.py`. This suite tests:

 1.  [_set_selected_variables()](https://github.com/ewengillies/tigramite/blob/58ac591c1cf88b08a10b22516547ec850f3ff168/tigramite/pcmci.py#L268)
     * Ensures default case is correct. 
     * Ensures can be set correctly.
     * Ensure throws errors when needed.
 2. [_set_sel_links()](https://github.com/ewengillies/tigramite/blob/58ac591c1cf88b08a10b22516547ec850f3ff168/tigramite/pcmci.py#L296)
    * Ensures default case is correct. 
    * Ensures can be set correctly.
    * Ensure throws errors when needed.
 3. [_iter_conditions()](https://github.com/ewengillies/tigramite/blob/9a639565434c76dad9296fac43ceb3a213f9bb96/tigramite/pcmci.py#L354)
    * Parent node is not in the conditions.
    * Ensures correct number of conditions are found in each iteration.
    * Ensures unique conditions sets are found in each iteration.
    * Ensure total number of conditions is correct.
    * `max_conds_dim`, `tau_min`, `tau_max` all effect the resulting conditions correctly.
 4. [_iter_indep_conds()](https://github.com/ewengillies/tigramite/blob/9a639565434c76dad9296fac43ceb3a213f9bb96/tigramite/pcmci.py#L1041)
    * Parent node is not in the conditions.
    * Maximum condition parameter is respected
    * Correct total number of links to test is recovered.
 5. [_sort_parents()](https://github.com/ewengillies/tigramite/blob/9a639565434c76dad9296fac43ceb3a213f9bb96/tigramite/pcmci.py#L379)
    * Parents correctly sorted by the test metric.
 6. [_check_tau_limits](https://github.com/ewengillies/tigramite/blob/9a639565434c76dad9296fac43ceb3a213f9bb96/tigramite/pcmci.py#L734) 
    * Correctly passes for good limits.
    * Correctly fails for bad limits.
 7. [get_corrected_pvalues](https://github.com/ewengillies/tigramite/blob/9a639565434c76dad9296fac43ceb3a213f9bb96/tigramite/pcmci.py#L1284)
    * Checks that `'none'` and `'fdr_bh'` are valid modes. 
    * Checks that `'none'` does nothing to the input.
    * Check that `exclude_contemporaneous` works
    * Check that auto-correlation elements are ignored.
 8. [return_significant_parents](https://github.com/ewengillies/tigramite/blob/9a639565434c76dad9296fac43ceb3a213f9bb96/tigramite/pcmci.py#L1328)
    * Correct number of links are returned for parents dictionary and link matrix.
    * Correct links are returned for parents dictionary and link matrix.

File: tigramite/pcmci.py [hunk 1]
```diff
@@ -1366,45 +1393,61 @@ def _print_significant_links(self,
...
+    def print_results(self, return_dict, alpha_level=0.05):
+        """Prints significant parents from output of MCI or PCMCI algorithms.
+
+        Parameters
+        ----------
+        return_dict : dict
+            Dictionary of return values, containing keys
+                * 'p_matrix'
+                * 'val_matrix'
+                * 'conf_matrix'
```
Comment (ewengillies) [inline tigramite/pcmci.py:1433]: conf_matrix should be optional as well

**Review by ewengillies — COMMENTED** [review]
fixed

**Comment by ewengillies** [comment]
Ready to be merged @jakobrunge

**Review by jakobrunge — APPROVED** [review]

**Pull request merged**